lib/streamlit/type_util.py

```diff
@@ -531,13 +532,23 @@ def pyarrow_table_to_bytes(table: pa.Table) -> bytes:
     return cast(bytes, sink.getvalue().to_pybytes())
 
 
-def convert_mixed_columns_to_string(
+def fix_unsupported_column_types(
     df: DataFrame, selected_columns: Optional[List[str]] = None
 ) -> DataFrame:
     for col in selected_columns or df.columns:
         column = df[col]
-        if column.dtype == "object" and "mixed" in infer_dtype(column):
+        if (
+            column.dtype == "object" and infer_dtype(column).startswith("mixed")
+        ) or column.dtype == "complex128":
             df[col] = column.astype(str)
+
+    # Fix the index in case it uses mixed type
+    if not selected_columns and (
+        not isinstance(df.index, MultiIndex)
+        and df.index.dtype == "object"
+        and infer_dtype(df.index).startswith("mixed")
```

Comment:
There aren't a lot of types returned by `infer_dtype` and the list seems to stay pretty stable (see [here](https://pandas.pydata.org/docs/reference/api/pandas.api.types.infer_dtype.html)). But I added the relevant mixed types explicitly, especially since `mixed-integer-float` actually seems to be handled fine by Arrow.